### PR TITLE
[flutter_tools] Use XDG_CONFIG_HOME dir by default for config files

### LINKED ISSUE #59430: Support XDG Base Directory Specfication
## Use case
There is a [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) which supports keeping home directory in Linux environments tidy. Currently, flutter keeps its global configuration in $HOME/.flutter and $HOME/.flutter_tool_state. According to the above mentioned specification, it is advised to keep configuration under $HOME/.config directory by default.

## Proposal
Allow user to set an environment variable to override default directory of flutter config files. This way it won't affect configuration of current users.

Alternatively, a fallback check to $HOME can be done if configuration does not exist in XDG base directory. This way XDG specification will be followed by default without having to set any environment variables.

### PR DESCRIPTION
**Pull request by Jesse-Bakker**
## Description

This PR changes the `Config` class in `flutter_tools` to use the XDG Base directory [specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) instead of putting files directly in the user's home directory. If those files are already present in the home directory, they are used instead.

## Related Issues

fixes #59430

## Tests

I added the following tests:

config_test.dart: `Config in home dir`, `Config in config dir`

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/config.dart [hunk 1]
```diff
@@ -21,10 +21,19 @@ class Config {
...
       _userHomePath(platform),
-      name,
+      '.flutter_$name'
     ));
+    if (!file.existsSync()) {
+      final FileSystemUtils fsUtils = FileSystemUtils(platform: platform, fileSystem: fileSystem);
+      file = fileSystem.file(fileSystem.path.join(
+        _userConfigPath(platform, fileSystem),
+        kConfigDir,
+        name)
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/config.dart:33]: nit: I would keep the closing `)`s on the same line

File: packages/flutter_tools/lib/src/base/config.dart [hunk 2]
```diff
@@ -21,10 +21,19 @@ class Config {
...
     @required Platform platform,
   }) {
-    final File file = fileSystem.file(fileSystem.path.join(
+    File file = fileSystem.file(fileSystem.path.join(
       _userHomePath(platform),
-      name,
+      '.flutter_$name'
     ));
+    if (!file.existsSync()) {
+      final FileSystemUtils fsUtils = FileSystemUtils(platform: platform, fileSystem: fileSystem);
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/config.dart:29]: Instead of using FSutils here I would just do:

```
file.parent.createSync(recursive: true);
```

The FSUtils will throw a fatal exception if the operation fails, which we don't want for the config/settings.

File: packages/flutter_tools/lib/src/base/config.dart [hunk 3]
```diff
@@ -53,8 +62,10 @@ class Config {
     }
   }
 
+  /// The default directory name for Flutter's configs
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/config.dart:65]: Please add some more context on how this is used. For example, talk about how if the config is written in home it is appended to the name, and if xdg then not.

File: packages/flutter_tools/lib/src/base/config.dart [hunk 4]
```diff
@@ -97,4 +108,12 @@ class Config {
       : 'HOME';
     return platform.environment[envKey] ?? '.';
   }
+
+  static String _userConfigPath(Platform platform, FileSystem filesystem) {
+    if (platform.operatingSystem == 'windows') {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/config.dart:113]: This seems wrong? Why no `if (!platform.isLinux) { ... }`

File: packages/flutter_tools/test/commands.shard/hermetic/downgrade_test.dart [hunk 5]
```diff
@@ -43,7 +43,7 @@ void main() {
   });
 
   testUsingContext('Downgrade exits on unknown channel', () async {
-    fileSystem.currentDirectory.childFile('.flutter_tool_state')
+    fileSystem.currentDirectory.childFile('tool_state')
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/hermetic/downgrade_test.dart:46]: Should this file name have stayed the same?

File: packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart [hunk 6]
```diff
@@ -335,7 +335,7 @@ void main() {
           ),
         ]);
         tempDir = globals.fs.systemTempDirectory.createTempSync('flutter_upgrade_test.');
-        flutterToolState = tempDir.childFile('.flutter_tool_state');
+        flutterToolState = tempDir.childFile('tool_state');
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart:338]: same here?

File: packages/flutter_tools/test/general.shard/config_test.dart [hunk 7]
```diff
@@ -81,4 +81,29 @@ void main() {
     expect(file.existsSync(), false);
     expect(bufferLogger.errorText, contains('Failed to decode preferences'));
   });
+
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/config_test.dart:84]:
```suggestion
```

File: packages/flutter_tools/test/general.shard/config_test.dart [hunk 8]
```diff
@@ -81,4 +81,29 @@ void main() {
...
+      'example',
+      fileSystem: memoryFileSystem,
+      logger: BufferLogger.test(),
+      platform: fakePlatform,
+    );
+    expect(config.getValue('hello'), 'bar');
+    expect(memoryFileSystem.file('.config/flutter/example').existsSync(), false);
+  });
+
+  testWithoutContext('Config in config dir', () {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/config_test.dart:98]: Longer comment here: config is created in the specified xdg directory if it does not already exist in the home dir?

**Review by jonahwilliams — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/base/config.dart (hunk 4 above)
Comment (Jesse-Bakker) [reply packages/flutter_tools/lib/src/base/config.dart:113]: You are right. I do however think `$XDG_CONFIG_HOME` should be the default on macOS as well, so it would be `if (!platform.isLinux && !platform.isMacOS)`.

File: packages/flutter_tools/test/commands.shard/hermetic/downgrade_test.dart (hunk 5 above)
Comment (Jesse-Bakker) [reply packages/flutter_tools/test/commands.shard/hermetic/downgrade_test.dart]: I changed `Config.test` to prepend the `.flutter_` prefix, so this can stay the same.

**Comment by Jesse-Bakker** [comment]
Realized that the previous version would not work correctly on windows. Fixed, refactored and fixed requested changes.

File: packages/flutter_tools/lib/src/base/config.dart [hunk 9]
```diff
@@ -13,18 +13,20 @@ import 'utils.dart';
 /// A class to abstract configuration files.
 class Config {
   /// Constructs a new [Config] object from a file called [name] in the
-  /// current user's home directory as determined by the [Platform] and
-  /// [FileSystem].
+  /// current user's configuration directory as determined by the [Platform]
+  ///and [FileSystem]. The configuration directory defaults to $XDG_CONFIG_HOME
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/config.dart:17]: good docs, but follow https://dart.dev/guides/language/effective-dart/documentation#do-start-doc-comments-with-a-single-sentence-summary and split off the additional details in a new paragraph.

File: packages/flutter_tools/lib/src/base/config.dart [hunk 10]
```diff
@@ -34,9 +36,12 @@ class Config {
     String name, {
     @required Directory directory,
     @required Logger logger,
-  }) => Config._(directory.childFile(name), logger);
+  }) =>
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/config.dart:39]: can you revert style changes from this PR that are unrelated to the change?

File: packages/flutter_tools/lib/src/base/config.dart [hunk 11]
```diff
@@ -92,9 +103,24 @@ class Config {
   //
   // Note that this is different from FileSystemUtils.homeDirPath.
   static String _userHomePath(Platform platform) {
-    final String envKey = platform.operatingSystem == 'windows'
-      ? 'APPDATA'
-      : 'HOME';
+    final String envKey =
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/config.dart:106]: while you are here (and if you're going to reformat it...), change this to `platform.isWindows` ? ... : ...

File: packages/flutter_tools/lib/src/base/config.dart [hunk 12]
```diff
@@ -92,9 +103,24 @@ class Config {
...
+  static String _configPath(
+      Platform platform, FileSystem fileSystem, String name) {
+    final String homeDirFile =
+        fileSystem.path.join(_userHomePath(platform), '.${kConfigDir}_$name');
+    if (!platform.isLinux && !platform.isMacOS) {
+      return homeDirFile;
+    }
+    if (fileSystem.isFileSync(homeDirFile)) {
+      return homeDirFile;
+    } else {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/config.dart:120]: you don't need an `else` block if you have an early return

File: packages/flutter_tools/lib/src/base/config.dart [hunk 13]
```diff
@@ -92,9 +103,24 @@ class Config {
...
+    final String envKey =
+        platform.operatingSystem == 'windows' ? 'APPDATA' : 'HOME';
     return platform.environment[envKey] ?? '.';
   }
+
+  static String _configPath(
+      Platform platform, FileSystem fileSystem, String name) {
+    final String homeDirFile =
+        fileSystem.path.join(_userHomePath(platform), '.${kConfigDir}_$name');
+    if (!platform.isLinux && !platform.isMacOS) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/config.dart:115]: think positive

File: packages/flutter_tools/lib/src/base/config.dart [hunk 14]
```diff
@@ -92,9 +103,24 @@ class Config {
...
+      return homeDirFile;
+    }
+    if (fileSystem.isFileSync(homeDirFile)) {
+      return homeDirFile;
+    } else {
+      final String configDir = platform.environment['XDG_CONFIG_HOME'] ??
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/config.dart:121]: pull out `XDG_CONFIG_HOME` into a constant and document it with a link to the specification

**Comment by jonahwilliams** [comment]
Can you also resolve the conflicting files?

**Review by jonahwilliams — APPROVED** [review]
LGTM!

Thank you for the contribution @Jesse-Bakker !
